### PR TITLE
TryFindSystemTimeZoneById

### LINKED ISSUE #66928: [API Proposal]: TryFindSystemTimeZoneById
### Background and motivation

The method for finding a timezone by a string, `TimeZoneInfo.FindSystemTimeZoneById`, throws a `TimeZoneNotFoundException` if it can't find one. AFAIK there isn't a non-throwing method in the runtime.

I use a version of the method below to [attempt to find a time zone or fallback to local time](https://github.com/danielchalmers/DesktopClock/blob/7ca9dc977223f1332bc88bd2fd7e3c3a7eb838a5/DesktopClock/SettingsHelper.cs#L11-L12):

```csharp
public static TimeZoneInfo GetTimeZone() =>
DateTimeUtil.TryGetTimeZoneById(Settings.Default.TimeZone, out var timeZoneInfo) ? timeZoneInfo : TimeZoneInfo.Local;
```

This is of course still possible using the current method but it's not as clean IMO.

### API Proposal

```csharp
namespace System
{
    public class TimeZoneInfo
    {
        public static bool TryFindSystemTimeZoneById(string id, out TimeZoneInfo timeZoneInfo)
        {
            try
            {
                timeZoneInfo = TimeZoneInfo.FindSystemTimeZoneById(id);
                return true;
            }
            catch (TimeZoneNotFoundException)
            {
                timeZoneInfo = null;
                return false;
            }
        }
    }
}
```

### API Usage

```csharp
// Ask user for timezone.
Console.WriteLine("Enter a timezone");
var tz = Console.ReadLine();

// Validate.
if(TimeZoneInfo.TryFindSystemTimeZoneById(tz, out var tzInfo))
    Console.WriteLine("Found timezone");
else
    Console.WriteLine("Invalid timezone");
```

```csharp
// Ask user for timezone.
Console.WriteLine("Enter a timezone");
var tz = Console.ReadLine();

// Try to find timezone or fall back to local time.
var finalTimeZone = TimeZoneInfo.TryFindTimeZoneById(tz, out var tzInfo) ? tzInfo : TimeZoneInfo.Local;
Console.WriteLine("Using " + finalTimeZone.DisplayName);
```

### Alternative Designs

_No response_

### Risks

_No response_

### PR DESCRIPTION
**Pull request by AlexRadch**
Close https://github.com/dotnet/runtime/issues/66928 issue.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.cs [hunk 1]
```diff
@@ -556,6 +557,44 @@ public static void ClearCachedData()
...
             ConvertTime(dateTime, FindSystemTimeZoneById(destinationTimeZoneId));
 
+        /// <summary>
+        /// Helper function for retrieving a TimeZoneInfo object by time_zone_name.
+        /// This function wraps the logic necessary to keep the private
+        /// SystemTimeZones cache in working order
+        ///
+        /// This function will either return a valid TimeZoneInfo instance or
+        /// it will throw 'InvalidTimeZoneException' / 'TimeZoneNotFoundException'.
+        /// </summary>
```
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.cs:567]: ></summary> [](http://example.com/codeflow?start=12&length=10)

Could you please add the param and return doc elements?

File: src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.cs [hunk 2]
```diff
@@ -556,6 +557,44 @@ public static void ClearCachedData()
...
+            {
+                throw new SecurityException(SR.Format(SR.Security_CannotReadFileData, id), e);
+            }
+            else
+            {
+                throw new TimeZoneNotFoundException(SR.Format(SR.TimeZoneNotFound_MissingData, id), e);
+            }
+        }
+
+        public static bool TryFindSystemTimeZoneById(string id, [NotNullWhenAttribute(true)] out TimeZoneInfo? timeZoneInfo)
```
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.cs:595]: > public static bool TryFindSystemTimeZoneById [](http://example.com/codeflow?start=7&length=45)

Please add the Xml doc elements to the new API.

File: src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Win32.cs [hunk 3]
```diff
@@ -320,48 +320,29 @@ private static TimeZoneInfo GetLocalTimeZoneFromWin32Data(in TIME_ZONE_INFORMATI
         /// This function will either return a valid TimeZoneInfo instance or
         /// it will throw 'InvalidTimeZoneException' / 'TimeZoneNotFoundException'.
         /// </summary>
-        public static TimeZoneInfo FindSystemTimeZoneById(string id)
+        private static TimeZoneInfoResult TryFindSystemTimeZoneById(string id, out TimeZoneInfo? timeZone, out Exception? e)
```
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Win32.cs:323]: >TryFindSystemTimeZoneById [](http://example.com/codeflow?start=42&length=25)

This is now almost exact to the same one on Unix file. I am seeing the only difference is the check `id.Length > MaxKeyLength` which we can apply on Unix version too. 

I suggest get rid of these two and consolidate them into the one you introduced in the common TimeZoneInfo.cs file.

File: src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.cs [hunk 4]
```diff
@@ -274,54 +274,34 @@ private static TimeZoneInfoResult TryGetTimeZoneFromLocalMachine(string id, out
...
-        public static TimeZoneInfo FindSystemTimeZoneById(string id)
+        private static TimeZoneInfoResult TryFindSystemTimeZoneById(string id, out TimeZoneInfo? timeZone, out Exception? e)
         {
             // Special case for Utc as it will not exist in the dictionary with the rest
             // of the system time zones.  There is no need to do this check for Local.Id
             // since Local is a real time zone that exists in the dictionary cache
             if (string.Equals(id, UtcId, StringComparison.OrdinalIgnoreCase))
             {
-                return Utc;
+                timeZone  = Utc;
```
Comment (danmoseley) [inline src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Unix.cs:287]:
```suggestion
                timeZone = Utc;
```

File: src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Win32.cs (hunk 3 above)
Comment (AlexRadch) [reply src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Win32.cs:323]: They make different calls:
`TryGetTimeZone(id, false, out timeZone, out e, cachedData);` -- Win32
`TryGetTimeZone(id, false, out timeZone, out e, cachedData, alwaysFallbackToLocalMachine: true);` -- Unix

File: src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.cs [hunk 5]
```diff
@@ -556,6 +557,58 @@ public static void ClearCachedData()
...
+            TimeZoneInfo? value;
+            Exception? e;
+
+            TimeZoneInfoResult result = TryFindSystemTimeZoneById(id, out value, out e);
+
+            if (result == TimeZoneInfoResult.Success)
+            {
+                return value!;
+            }
+            else if (result == TimeZoneInfoResult.InvalidTimeZoneException)
```
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.cs:582]: >else [](http://example.com/codeflow?start=12&length=4)

this `else` is not needed.

File: src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Win32.cs (hunk 3 above)
Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Win32.cs:325]: Still worth moving the common code to timezoneinfo.cs and keep here only the different code. what do you think?

File: src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.cs [hunk 6]
```diff
@@ -556,6 +557,90 @@
...
+            // since Local is a real time zone that exists in the dictionary cache
+            if (string.Equals(id, UtcId, StringComparison.OrdinalIgnoreCase))
+            {
+                timeZone = Utc;
+                e = default;
+                return TimeZoneInfoResult.Success;
+            }
+
+            ArgumentNullException.ThrowIfNull(id);
+            if (id.Length == 0 || id.Length > MaxKeyLength || id.Contains('\0'))
```
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.cs:629]: >MaxKeyLength [](http://example.com/codeflow?start=46&length=12)

this is giving build error. Please try to compile and run the tests before updating. Thanks!

**Comment by tarekgh** [comment]
>         private const int MaxKeyLength = 255;

move it to TimeZoneInfo.cs

---
Refers to: src/libraries/System.Private.CoreLib/src/System/TimeZoneInfo.Win32.cs:35 in 2899919. [](commit_id = 2899919b1fe40f51c8936b7e0dd7ff0d356c9dc9, deletion_comment = False)

**Review by tarekgh — APPROVED** [review]

**Comment by tarekgh** [comment]
Thanks @AlexRadch for providing these changes.

**Pull request merged**